### ISSUE TITLE
These reports and report filters needed on Requests Reports page.

### ISSUE THREAD

**Issue by kfogel** [redcross/smoke-alarm-portal#82]
This is the promised ticket describing the reports and report filters we need.  They're pretty simple, I think -- hopefully drywall gets us most of this for free or nearly-free.

Not going into great detail about UI here because the default UI (based on the currently available "all requests" report) seems pretty good already, so any UI improvements are probably best done iteratively after we get the basic reports up.

Here's the list:

* [x] We need a report of requests filtered by date range
* [x] We need a report of requests filtered by Red Cross Region.  The region names as we should display them come from the `region_alt_name` field in the `config/recipients.json.tmpl` file that is currently on the `email_demo` branch (see how that field is used in `routes/index.js` on that branch).  I will be merging `email_demo` branch into `master` soon, of course, like today.  However, this item shouldn't be blocked on that merge, since the DB already has the region mappings; it's just that after the merge, the code that implements this item might get tweaked a bit (and if so I'll take care of that tweak as part of the merge).  By the way, I think we don't currently cache the RC Region in the request record when we make a new request entry in the DB.  In theory, we can always re-derive the RC Region by doing a query involving the request's zip code; in practice, if you want to start caching the RC Region in the request's record at the time we receive the request from the front-page form, I think that's fine :-).
* [x] Obviously, the combination of the above two: report of requests by RC Region and filtered by date.
* [x] We'll get pagination automatically, right?  Can we offer a "500 results per page" option too, for the pagination?  (I think right now the max is 100.)
* [ ] Do we get CSV export of any result set for free, or do we have to build that?  (I assume that exports are of the full result set, i.e., are not limited by pagination.)
* [x] Any request result set should be sortable by date, and by default result lists should display in earliest-to-latest order.
* [x] All requests should show a column for the date the request was received.  This format is fine: "YYYY-MM-DD HH:MM" (with 24-hour clock for HH)
* [x] All requests should show a column for unique ID (this is already filed as issue #81; I'm just including here so you have everything in one place -- feel free to either remove it from here or to close issue #81 as a dup of this, if you want)

Let me know if any of these look like they might be complicated, so that we can start thinking of workarounds or simplifications now, and so we can prioritize properly.

**Comment by trystant** [comment]
I'm running into an error with the results from the query getting the paged results, so I'm currently figuring out the source of this problem. a ![screenshot](http://grandkru.com/images/paged_result_retrieval_error.png) is here. The error message is: 

````Uncaught TypeError: Cannot read property 'total' of undefined````

I am figuring out the source of the problem.

**Comment by cecilia-donnelly** [comment]
@trystant had been implementing pagination with 500 results per page, but ran into a more global pagination problem.  This had been hinted at here https://github.com/OpenTechStrategies/smoke-alarm-portal/issues/44#issuecomment-124738737, but he has now found the source of the error.

> It seems that the results were not modified to use Sequelize across all parts of the admin.  The quick and effective way to resolve this is to replace the DB async query with a promise and stick that into the existing async framework, which is how I got the individual request page and edit (on my machine) to work.

**Comment by trystant** [comment]
Hi all,

As noted in a343e05; the pagination, id and date columns are in place. The export is not free, unfortunately. We'd need to add some code in order to do the export. I think that converting the json to csv and setting the application type should be sufficient.

What's left is the sorting and filtering of the results. Both of these will be accomplished by placing the proper options in the query which is made. I need to write the sequelize version of the compound query and then those will be in place. I will tackle this tomorrow. 

Thanks,
Mark

**Comment by trystant** [comment]
@kfogel, @cecilia-donnelly,  when you have a moment, I'd like you to check out what I've done in this new branch, 82_integrated_refactored_save_request. Due to the challenges with working with the code while getting region update working, I refactored it into a way that I believe makes working with this simpler.

I've tested this, and all the functionality works, with the exception of sending an email. For some reason, mailgun returns an error where the body is undefined. I've got the [gist](https://gist.github.com/trystant/f39c4895174467564be6) of it here.

Once that gets resolved, I wanted to confirm whether we could get rid of routes/index.js and have all this functionality in views/index.js.

**Comment by trystant** [comment]
In order to have the region saved properly and our collective sanity, views/index.js was refactored to handle promise chains better and separate pieces of functionality, which allowed me to add the query to update the region relatively easily. 

As I said in the email, I used a migration to change the column name, so remember to `sequelize db:migrate` before you `npm start`

**Comment by trystant** [comment]
After thinking things through, I've decided to go with displaying the region name from the recipients table and returning that in the query to the front end. I'll be modifying the results of the findAll in `views/admin/requests/index.js`. 

After that, the last change is implementing the region filter in `views/admin/requests/index.js` and ``views/admin/requests/index.jade`.

**Comment by kfogel** [comment]
Thanks, @trystant.

Background: one poor design decision I made was in how `config/recipients.json` labels the two versions of the names for the Regions.  The top-level keys in the file are region names, obviously, and then there's the `region_alt_name` field -- well, great: it's totally unclear from that which one is the one that should be displayed to the user!  That is, which one is the official Red Cross Region name that (for example) a Red Cross administrator logging in to the admin UI should see, or (for another example) the one that should be used in the body of the email sent to the regional contact when a request comes in.  My apologies for the lack of clarity in that part of the code; it makes more mental work for everyone.

The resolution: I can see in the code that we have this line:

    var regionPresentableName = recipients_table[selectedRegion.region]["region_alt_name"];

Whew, at least I chose a variable name that communicates something :-).  So, clearly, `region_alt_name` is the one that should be displayed.

Thus, probably we should do two things:

1. Change the top-level keys in `config/recipients.json` to something that is obviously not meant for display, e.g., "Iowa" would become "rc_region_iowa", and "Idaho and Montana" would become "rc_region_idaho_montana", etc.
2. Change the name of the `region_alt_name` field to `region_display_name`.

Then the distinction between the two would be completely clear.  One is for use in code as a unique identifier for that region; the other is the text you put on the screen (or in an email) when you want to communicate that region name to users.

Would the above changes clarify the question/decision referred to in your comment above?

**Comment by trystant** [comment]
@kfogel

I believe that that does help, and thanks for that. I will make those
changes in conjunction with the changes I outlined earlier in this thread.
I am going to have the display name returned as the last property of each
result row, then put the filter pieces in place.

I have a meeting now, and will be in IRC in 2 hours and will submit a
commit with this work shortly after. Thanks



On Thu, Aug 6, 2015 at 11:43 AM, Karl Fogel <notifications@github.com>
wrote:

> Thanks, @trystant <https://github.com/trystant>.
>
> Background: one poor design decision I made was in how
> config/recipients.json labels the two versions of the names for the
> Regions. The top-level keys in the file are region names, obviously, and
> then there's the region_alt_name field -- well, great: it's totally
> unclear from that which one is the one that should be displayed to the
> user! That is, which one is the official Red Cross Region name that (for
> example) a Red Cross administrator logging in to the admin UI should see,
> or (for another example) the one that should be used in the body of the
> email sent to the regional contact when a request comes in. My apologies
> for the lack of clarity in that part of the code; it makes more mental work
> for everyone.
>
> The resolution: I can see in the code that we have this line:
>
> var regionPresentableName = recipients_table[selectedRegion.region]["region_alt_name"];
>
> Whew, at least I chose a variable name that communicates something :-).
> So, clearly, region_alt_name is the one that should be displayed.
>
> Thus, probably we should do two things:
>
>    1. Change the top-level keys in config/recipients.json to something
>    that is obviously not meant for display, e.g., "Iowa" would become
>    "rc_region_iowa", and "Idaho and Montana" would become
>    "rc_region_idaho_montana", etc.
>    2. Change the name of the region_alt_name field to region_display_name.
>
> Then the distinction between the two would be completely clear. One is for
> use in code as a unique identifier for that region; the other is the text
> you put on the screen (or in an email) when you want to communicate that
> region name to users.
>
> Would the above changes clarify the question/decision referred to in your
> comment above?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/redcross/smoke-alarm-portal/issues/82#issuecomment-128414554>
> .
>



-- 

Mark Nyonmark@grandkru.com

blog.grandkru.com

github.com/trystant

**Comment by kfogel** [comment]
(Please let me know if this issue isn't the right place to ask this.)

Porting this question over from issue #99, as I think it's more appropriately asked here:

When we store the RC Region associated with a given request, why are we storing the region's display name rather than its internal code name?  The whole point of internal identifiers is to be the thing that's stored in the database.  This is exactly the same phenomenon as why, say, a User record (in a system that has user records) has a unique identifying user-number associated with it (a primary key for that user, basically), so that if the user's name changes -- say, to add a "Ms." or to fix a typo or whatever -- that doesn't change the identity of the underlying record.

We should be treating Red Cross regions the same way.  The display name is for display; it's not the unique identifier that we toss around internally.  These display names *have* changed in the recent past, by the way -- in fact, for the very region we're doing this project with: the region "Greater Chicago" became "Chicago and Northern Illinois".  If we use the display name as the internal key, then the next time this happens, the system will suddenly start displaying the wrong name (the region itself didn't change, just the name) -- but worse, *searches* for requests in a region with a recently changed name will fail to return results that they should return.

In other words, Badness Is About To Ensue :-).  Let's fix this.

**Comment by kfogel** [comment]
(And if I've just misunderstood what the code is doing, then my apologies!  The above analysis was from a slightly-more-than-cursory but somewhat-less-than-deep-dive look at the code in question.)

**Comment by cecilia-donnelly** [comment]
The switch from storing the displayed region name to a region code has broken the sort by region functionality.

This url: `http://localhost:3000/admin/requests#q/search=&region=Minnesota` correctly returns requests in the Minnesota region.  However, when I choose Minnesota from the dropdown, *this* url is returned: `http://localhost:3000/admin/requests#q/search=&region=rc_minnesota`.  You can see that the code name is being passed in the search, but evidently it is not what is being stored in the db, since no requests are returned by that search.

**Comment by cecilia-donnelly** [comment]
After recreating my db, this is fixed, so I'm going to close this issue.
